src/x11.rs

```diff
@@ -187,138 +207,205 @@ impl X11Impl {
             .swbuf_err("Failed to get geometry reply")?;
 
         // See if SHM is available.
-        let shm_info = {
-            let present = display.is_shm_available;
-
-            if present {
-                // SHM is available.
-                Some(ShmInfo {
-                    seg: None,
-                    done_processing: None,
-                })
-            } else {
-                None
-            }
+        let buffer = if display.is_shm_available {
+            // SHM is available.
+            Buffer::Shm(ShmBuffer {
+                seg: None,
+                done_processing: None,
+            })
+        } else {
+            // SHM is not available.
+            Buffer::Wire(Vec::new())
         };
 
         Ok(Self {
             display,
             window,
             gc,
             depth: geometry_reply.depth,
-            shm: shm_info,
+            buffer,
+            width: 0,
+            height: 0,
         })
     }
 
-    pub(crate) unsafe fn set_buffer(&mut self, buffer: &[u32], width: u16, height: u16) {
-        // Draw the image to the buffer.
-        let result = unsafe { self.set_buffer_shm(buffer, width, height) }.and_then(|had_shm| {
-            if had_shm {
-                Ok(())
-            } else {
-                log::debug!("Falling back to non-SHM method");
-                self.set_buffer_fallback(buffer, width, height)
-            }
-        });
-
-        if let Err(e) = result {
-            log::error!("Failed to draw image to window: {}", e);
+    /// Resize the internal buffer to the given width and height.
+    pub(crate) fn resize(&mut self, width: u32, height: u32) -> Result<(), SoftBufferError> {
+        log::trace!(
+            "resize: window={:X}, size={}x{}",
+            self.window,
+            width,
+            height
+        );
+
+        // Width and height should fit in u16.
+        let width: u16 = width
+            .try_into()
+            .or(Err(SoftBufferError::SizeOutOfRange { width, height }))?;
+        let height: u16 = height.try_into().or(Err(SoftBufferError::SizeOutOfRange {
+            width: width.into(),
+            height,
+        }))?;
+
+        if width != self.width || height != self.height {
+            self.buffer
+                .resize(&self.display.connection, width, height)
+                .map_err(|err| {
```

Comment:
The `swbuf_err` method expresses this pattern more succinctly.

Reply:
Done.

We'll want to think about how best to define the error type (or use multiple error types?) and move any helpers like this to `error.rs` and use them in all the backends. But that can be done later.

Reply:
Opened https://github.com/rust-windowing/softbuffer/issues/73.